### PR TITLE
Implement new feature GetMaxTemperature

### PR DESCRIPTION
**Pull request by Xodoc**

### CODE REVIEW

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs [hunk 1]
```diff
@@ -64,6 +67,50 @@ public async Task<WeatherForecast> GetWeatherForecastAsync(string cityName, int
...
+                        maxTemp.CityName = weather.Name;
+                        maxTemp.Temp = weather.Main.Temp;
+                        maxTemp.CountSuccessfullRequests++;
+                        maxTemp.RunTime = stopwatch.ElapsedMilliseconds;
+                    }
+
+                    temperatures.Add(maxTemp);
+                    x++;
+
+                    Monitor.Pulse(locker);
```
Comment (denja2) [inline ConsoleApp/DAL/Repositories/WeatherRepository.cs:107]: P1 remove this

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs [hunk 2]
```diff
@@ -64,6 +67,50 @@ public async Task<WeatherForecast> GetWeatherForecastAsync(string cityName, int
...
+            var temperatures = new ConcurrentBag<MaxTemperature>();
+            var locker = new object();
+            var amountCities = cityNames.Count;
+            var x = 0;
+
+            Parallel.For(0, amountCities, i =>
+            {
+                var stopwatch = new Stopwatch();
+
+                lock (locker)
```
Comment (denja2) [inline ConsoleApp/DAL/Repositories/WeatherRepository.cs:81]: P1 Check if it necessary

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs [hunk 3]
```diff
@@ -64,6 +67,50 @@ public async Task<WeatherForecast> GetWeatherForecastAsync(string cityName, int
...
         }
 
+        public List<MaxTemperature> GetTemperatures(List<string> cityNames)
+        {
+            var temperatures = new ConcurrentBag<MaxTemperature>();
+            var locker = new object();
+            var amountCities = cityNames.Count;
+            var x = 0;
+
+            Parallel.For(0, amountCities, i =>
```
Comment (denja2) [inline ConsoleApp/DAL/Repositories/WeatherRepository.cs:77]: P2 foreach

File: ConsoleApp/ConsoleApp/Commands/GetMaxTemperatureCommand.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        public string Title => "Show max temperature among cities";
+
+        public GetMaxTemperatureCommand(IWeatherService weatherService)
+        {
+            _weatherService = weatherService;
+        }
+
+        public async Task Execute()
+        {
+            Console.Write("Input number of cities: ");
```
Comment (denja2) [inline ConsoleApp/ConsoleApp/Commands/GetMaxTemperatureCommand.cs:22]: P1 array of cities

**Review by denja2 — CHANGES_REQUESTED** [review]

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs [hunk 5]
```diff
@@ -64,6 +67,50 @@ public async Task<WeatherForecast> GetWeatherForecastAsync(string cityName, int
             return result;
         }
 
+        public List<MaxTemperature> GetTemperatures(List<string> cityNames)
+        {
+            var temperatures = new ConcurrentBag<MaxTemperature>();
```
[the remark this answers is not in the document]
Comment (Xodoc) [reply ConsoleApp/DAL/Repositories/WeatherRepository.cs:72]: done

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs (hunk 1 above)
Comment (Xodoc) [reply ConsoleApp/DAL/Repositories/WeatherRepository.cs:107]: done

File: ConsoleApp/BL/Interfaces/IWeatherService.cs [hunk 6]
```diff
@@ -9,5 +9,7 @@ public interface IWeatherService
         Task<string> GetWeatherAsync(string cityName);
 
         Task<string> GetWeatherForecastAsync(string cityName, int days);
+
+        Task<string> GetMaxTemperatureAsync(List<string> cityNames, bool debugInfo);
```
[the remark this answers is not in the document]
Comment (Xodoc) [reply ConsoleApp/BL/Interfaces/IWeatherService.cs:13]: done

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs (hunk 2 above)
Comment (Xodoc) [reply ConsoleApp/DAL/Repositories/WeatherRepository.cs:81]: done

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs (hunk 3 above)
Comment (Xodoc) [reply ConsoleApp/DAL/Repositories/WeatherRepository.cs:77]: done

File: ConsoleApp/BL/Services/WeatherService.cs [hunk 7]
```diff
@@ -43,12 +46,53 @@ public async Task<string> GetWeatherForecastAsync(string cityName, int days)
...
+
+            var responseMessage = "";
+
+            if (_config.IsDebug == true)
+            {
+                responseMessage += $"[Debug]\n";
+
+                foreach (var temp in maxTemps)
+                {
+                    if (temp.CountFailedRequests > 0)
```
Comment (denja2) [inline ConsoleApp/BL/Services/WeatherService.cs:69]: P3 Is failed

File: ConsoleApp/BL/Services/WeatherService.cs [hunk 8]
```diff
@@ -43,12 +46,53 @@ public async Task<string> GetWeatherForecastAsync(string cityName, int days)
...
             return responseMessage;
         }
 
+        public Task<string> GetMaxTemperatureAsync(List<string> cityNames)
+        {
+            _validator.ValidateCityNames(cityNames);
+
+            var maxTemps = _weatherRepository.GetTemperatures(cityNames);
+
+            var responseMessage = "";
```
Comment (denja2) [inline ConsoleApp/BL/Services/WeatherService.cs:61]: P3 stringBuilder

File: ConsoleApp/BL/Services/WeatherService.cs [hunk 9]
```diff
@@ -43,12 +46,53 @@ public async Task<string> GetWeatherForecastAsync(string cityName, int days)
...
+                    if (temp.CountFailedRequests > 0)
+                        responseMessage += $"City: {temp.CityName}. Error: Invalid input data. Timer: {temp.RunTime} ms.\n";
+                    else
+                        responseMessage += $"City: {temp.CityName}. Temperature: {temp.Temp}°C. Timer: {temp.RunTime} ms.\n";
+                }
+            }
+
+            var maxTemp = CalculateTotalsForMessage(maxTemps);
+
+            responseMessage += $"City with the highest temperature {maxTemp.Temp}°C: {maxTemp.CityName}." +
```
Comment (denja2) [inline ConsoleApp/BL/Services/WeatherService.cs:78]: P3 use interpolation

File: ConsoleApp/ConsoleApp/Commands/GetMaxTemperatureCommand.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        public GetMaxTemperatureCommand(IWeatherService weatherService)
+        {
+            _weatherService = weatherService;
+        }
+
+        public async Task Execute()
+        {
+            Console.Write("Input city names: ");
+            var cities = Console.ReadLine();
+            var array = cities.Split(new string[] {" "}, StringSplitOptions.RemoveEmptyEntries);
```
Comment (denja2) [inline ConsoleApp/ConsoleApp/Commands/GetMaxTemperatureCommand.cs:24]: P2 notices from demo. P3 move splitting to a separate class

File: ConsoleApp/BL/Services/WeatherService.cs [hunk 11]
```diff
@@ -43,12 +46,53 @@ public async Task<string> GetWeatherForecastAsync(string cityName, int days)
...
 
-            weatherForecastDtos.ForEach(x => responseMessage += string.Join(",",
-                $"{x.CityName} weather forecast:\nDay {numberOfDay++}: {x.Temp}°C. {x.Description}\n"));
+            weatherForecastDtos.ForEach(x => responseMessage +=
+                $"{x.CityName} weather forecast:\nDay {numberOfDay++}: {x.Temp}°C. {x.Description}\n");
 
             return responseMessage;
         }
 
+        public Task<string> GetMaxTemperatureAsync(List<string> cityNames)
```
Comment (denja2) [inline ConsoleApp/BL/Services/WeatherService.cs:55]: P2 IEnumerable

File: ConsoleApp/BL/Validators/Validator.cs [hunk 12]
```diff
@@ -30,5 +31,13 @@ public void ValidateModel(string cityName, int days)
                 throw new ValidatorException("\nInvalid data entered");
             }
         }
+
+        public void ValidateCityNames(List<string> cityNames)
```
Comment (denja2) [inline ConsoleApp/BL/Validators/Validator.cs:35]: P2 Ienumeruble

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs [hunk 13]
```diff
@@ -64,6 +66,37 @@ public async Task<WeatherForecast> GetWeatherForecastAsync(string cityName, int
             return result;
         }
 
+        public List<TemperatureInfo> GetTemperatures(List<string> cityNames)
+        {
+            var temperatures = new ConcurrentBag<TemperatureInfo>();
+            var amountCities = cityNames.Count;
+
+            Parallel.ForEach(cityNames, name =>
```
Comment (denja2) [inline ConsoleApp/DAL/Repositories/WeatherRepository.cs:74]: P2 Async

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs [hunk 14]
```diff
@@ -64,6 +66,37 @@ public async Task<WeatherForecast> GetWeatherForecastAsync(string cityName, int
...
+            {
+                var stopwatch = new Stopwatch();
+                var weather = new CurrentWeather();
+
+                if (_config.IsDebug)
```
Comment (denja2) [inline ConsoleApp/DAL/Repositories/WeatherRepository.cs:79]: p3 Manage debug

**Review by denja2 — CHANGES_REQUESTED** [review]

File: ConsoleApp/BL/Services/WeatherService.cs (hunk 7 above)
Comment (Xodoc) [reply ConsoleApp/BL/Services/WeatherService.cs:69]: done

File: ConsoleApp/BL/Services/WeatherService.cs (hunk 11 above)
Comment (Xodoc) [reply ConsoleApp/BL/Services/WeatherService.cs:55]: done

File: ConsoleApp/BL/Validators/Validator.cs (hunk 12 above)
Comment (Xodoc) [reply ConsoleApp/BL/Validators/Validator.cs:35]: done

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs (hunk 13 above)
Comment (Xodoc) [reply ConsoleApp/DAL/Repositories/WeatherRepository.cs:74]: done

File: ConsoleApp/DAL/Repositories/WeatherRepository.cs (hunk 14 above)
Comment (Xodoc) [reply ConsoleApp/DAL/Repositories/WeatherRepository.cs:79]: done

File: ConsoleApp/ConsoleApp/Commands/GetMaxTemperatureCommand.cs (hunk 10 above)
Comment (Xodoc) [reply ConsoleApp/ConsoleApp/Commands/GetMaxTemperatureCommand.cs:24]: done

File: ConsoleApp/BL/Services/WeatherService.cs (hunk 8 above)
Comment (Xodoc) [reply ConsoleApp/BL/Services/WeatherService.cs:61]: done

File: ConsoleApp/ConsoleApp/Commands/GetMaxTemperatureCommand.cs [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            _weatherService = weatherService;
+        }
+
+        public async Task Execute()
+        {
+            Console.Write("Input city names: ");
+            var cities = Console.ReadLine();
+
+            var cityNames = new StringSplit(cities).SplitNames();
```
Comment (denja2) [inline ConsoleApp/ConsoleApp/Commands/GetMaxTemperatureCommand.cs:25]: P2 Inject in constructor; pass parameters 'cities' in method not class; class name StringSplitter

File: ConsoleApp/ConsoleApp/Commands/HelperClasses/StringSplit.cs [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+        private readonly string _cityNames;
+
+        public StringSplit(string cityNames)
+        {
+            _cityNames = cityNames;
+        }
+
+        public string[] SplitNames() 
+        {
+            return _cityNames.Split(new string[] { " ", "," }, StringSplitOptions.RemoveEmptyEntries);
```
Comment (denja2) [inline ConsoleApp/ConsoleApp/Commands/HelperClasses/StringSplit.cs:16]: Trim spaces

File: ConsoleApp/ConsoleApp/Commands/GetMaxTemperatureCommand.cs (hunk 15 above)
Comment (Xodoc) [reply ConsoleApp/ConsoleApp/Commands/GetMaxTemperatureCommand.cs:25]: done

File: ConsoleApp/ConsoleApp/Commands/HelperClasses/StringSplit.cs (hunk 16 above)
Comment (Xodoc) [reply ConsoleApp/ConsoleApp/Commands/HelperClasses/StringSplit.cs:16]: done

File: ConsoleApp/BL/Services/WeatherService.cs (hunk 9 above)
Comment (Xodoc) [reply ConsoleApp/BL/Services/WeatherService.cs]: done

**Pull request merged**